x-pack/plugins/enterprise_search/public/applications/app_search/components/engine/engine_nav.tsx

```diff
@@ -99,12 +99,12 @@ export const EngineNav: React.FC = () => {
           )}
         </EuiText>
       </SideNavItem>
-      <SideNavLink to={engineRoute} data-test-subj="EngineOverviewLink">
+      <SideNavLink to={generatePath(ENGINE_PATH)} data-test-subj="EngineOverviewLink">
```

Comment:
I don't think I like this part of this PR.

I'm not convinced that it's a good idea to couple this with a Logic.

```
const { generatePath } = useValues(EngineLogic);
<a href={generatePath(ENGINE_PATH, { engineName: 'another-engine' })}>Go to another engine</a>
```

I see you posted a solution to handling routes to *other* engines. At this point, if you're doing this, the `generatePath` selector is completely unrelated to anything in state, and doesn't really feel like it belongs.

And if you're doing this...
```
generatePath(SOME_EXAMPLE_PATH, { exampleId: 12345 })}
```

Half of the params are populate from state, and half from what you pass in. Which again just makes me scratch my head a bit when I see it.

My high level take on this is it muddies the water a bit and adds some complexity. I think I prefer the explicitness of having this disconnected from a logic. I do really like using a `generatePath` helper though, just not connected to Logic or state.

Reply:
I deliberately named it after the generatePath helper to associate it with RR's, because in my head "it's the exact same function, but with `engineName` filled in for you from state".

If folks think it's super confusing at a glance, I can rename it to `generateEnginePath` for extra clarity for sure.

> Half of the params are populated from state, and half from what you pass in. Which again just makes me scratch my head a bit when I see it.

I'm not sure why. That's not really different from other helpers that we use - for example, [KibanaLogic.values.navigateToUrl](https://github.com/elastic/kibana/blob/master/x-pack/plugins/enterprise_search/public/applications/shared/kibana/kibana_logic.ts#L39-L41) which essentially is an extra layer between Kibana's native `navigateToUrl` but with extra helpers and dependencies pre-filled, so we're not constantly passing around `history` and `http` dependencies into that function.

One of the whole reasons of using redux/Kea is to reduce is the amount of variables you're passing around endlessly when you can just fetch it once from state.

> the generatePath selector is completely unrelated to anything in state

I'm good in those scenarios with just importing the default React Router `generatePath` in that case. Probably should have done that o start, I can definitely make that change.

Reply:
I've made the above changes (`EngineLogic.values.generatePath` -> `EngineLogic.values.generateEnginePath` and also changed Result and EnginesTable to use RR's native `generatePath`). Take a look at the new diff and let me know what you think.

I want to be very clear about something however:

> I do really like using a generatePath helper though, just not connected to Logic or state.

I will not merge this PR in if the helper does not auto-fill the `engineName` param for you. I would strongly rather close it and continue on with our routes as-is. Not having to get and pass `engineName` around endlessly was the point of this helper, and I disagree that it is muddying anything (from my POV, it's a dev QOL improvement), but if we can't agree on that point, then there's no point to this PR in my opinion.

From my POV, these are equivalent in annoyingness:
```js
import { generateEngineRoute } from '../../routes';
const { engineName } = useValues(EngineLogic);
<a href={getEngineRoute(engineName) + SOME_ENGINE_PATH}>Some engine link</a>

import { generateSomeRoute } from '../../routes';
const { engineName } = useValues(EngineLogic)
<a href={generateSomeRoute(engineName)}>Some link</a>
```
1. You still have to import EngineLogic and then destructure engineName
2. You then have to pass engineName still, endlessly, for 90% of our routing
3. With all due respect to Byron who I know dislikes concatenating, but how is losing a `+` that significant of dev QOL improvement over not having to unnecessarily pass `engineName`?
4. Additionally, I'm not a fan (and want to get rid of) all of our `getXYZRoute` helpers. Those feel unnecessary and scale poorly (if we add infinite routes, we're adding an infinite +1 of `getXYZ` routes which feel like an unnecessary import over just importing the actual path and passing it to a generatePath helper).

Reply:
So this is the difference between the two options, right?

```
import { ENGINE_DOCUMENTS_PATH } from '../../routes';

### With this helper
const { generateEnginePath } = useValues(EngineLogic);
generateEnginePath(ENGINE_DOCUMENTS_PATH);
```

```
### Without this helper
import { generatePath } from 'react-router-dom'
import { ENGINE_DOCUMENTS_PATH } from '../../routes';
...
const { engineName } = useValues(EngineLogic);
generatePath(ENGINE_DOCUMENTS_PATH, { engineName });
```

You're right, it is slightly less code.

The tradeoff is now there are multiple ways of generating paths in the code. It's possible to get yourself into a situation where you have multiple styles in the same file...

```
import { ENGINE_DOCUMENTS_PATH, SETTINGS_PATH} from '../../routes';
import { generatePath } from 'react-router-dom';

const { generateEnginePath } = useValues(EngineLogic);

generateEnginePath(ENGINE_DOCUMENTS_PATH);
generatePath(SETTINGS_PATH)
```

That's my only point. It's not a deal breaker. I also don't care that much. Feel free to merge this PR.

Reply:
> The tradeoff is now there are multiple ways of generating paths in the code. It's possible to get yourself into a situation where you have multiple styles in the same file...

Heh, I'm chuckling a bit because was actually the intention of my original helper, which was to essentially replace RR's base `generatePath` completely and not potentially have two copies floating around. The idea is that `EngineLogic.values.generatePath` does everything `generatePath` does, just with `engineName` autofilled (since that's such a large part of our app), and the 2nd arg can be passed in & overridden super easily.

To be honest though, it wouldn't super bother me if someone was using both `generatePath` and `generateEnginePath` in the same file. I like that the helpers are flexible and have the same understandable API/usage & output, and that there's a distinction in the name now.

What I **would** like to push for avoiding in the future is having all those `getXYZRoute` fns in our old `routePaths.ts` file (e.g. `getQueryDetailRoute`, `getCrawlerRecentCrawlsRoute`) which were simply super shallow wrappers around `generatePath`. Just IMO, `generatePath` is flexible enough as it is without needing to redefine extra specific layers on top of it, and now with the `generateEnginePath` helper, I think just importing the route directly is simpler:

```jsx
// BEFORE
import { getReindexJobRoute } from '../../routes';
const { engineName } = useValues(EngineLogic);
getReindexJobRoute(engineName, 12345);


// AFTER
import { ENGINE_REINDEX_JOB_PATH } from '../../routes';
const { generateEnginePath } = useValues(EngineLogic);
generateEnginePath(ENGINE_REINDEX_JOB_PATH, { activeReindexJobId: 12345 });
```

To me, the 2nd method is clearer as to what you're dealing with & what param you're passing (especially w/o `engineName` just being arbitrarily present). And we'll also be saving ourselves from having to tediously define new helper functions with every new route we add - just a single reusable fn is more scalable and easier to follow (just IMO).

Reply:
> What I would like to push for avoiding in the future is having all those getXYZRoute fns in our old routePaths.ts file (e.g. getQueryDetailRoute, getCrawlerRecentCrawlsRoute) which were simply super shallow wrappers around generatePath

Yes, I definitely agree.